bridges/snowbridge/docs/v2.md

```diff
@@ -0,0 +1,310 @@
+3. The relayer dry-runs m on Ethereum to decide whether the message is profitable to deliver.
+4. The relayer finally delivers the message together with a relayer-controlled address $u$ on AH where the relayer reward will be lazily deposited.
+
+## Step 7: Relayer delivers proof of delivery to BH
+
+The proof of delivery is essentially a merkle proof for the `InboundMessageAccepted` event log.
+
+When BH processes the proof of delivery:
+
+1. The command $m$ is removed from storage items $M$ and $P$.
+2. The relayer reward is tracked in storage $R$, where $R(u)$ is the accumulated rewards that can be claimed by account $u$.
+
+# Ethereum→Polkadot
+
+## Step 1: Submit send on Gateway
+
+The interface that the Gateway will use to initiate transfers will be similar to the interface from `transfer_assets_using_type_and_then` extrinsic that we currently use to initiate transfers from the Polkadot to Ethereum direction.
+
+1. It must allow multiple assets to be transferred and specify the transfer type: Local, Destination or Teleport asset transfer types. It is the job of the User Agent/UX layer to fill in this information correctly.
+2. It must allow specifying a destination which is `Address32` , `Address20` or a custom scale-encoded XCM payload that is executed on the destination. This is how we will support `Transact` , the User Agent/UX layer can build a scale-encoded payload with an encoded transact call.
+3. The same interface is used for both PNA (Polkadot Assets) and ERC20 tokens. Internally we will still look up whether the token is registered as a PNA or ERC20 for the purpose of minting/locking burning/unlocking logic. The asset transfer type chosen by the UX layer will inform the XCM that is built from the message on BH.
+
+```solidity
+enum Kind {
+    Index,
+    Address32,
+    Address20,
+    XCMPayload,
+}
+
+struct Beneficiary {
+    Kind kind;
+    bytes data;
+}
+
+enum AssetTransferType {
+	ReserveDeposit, ReserveWithdraw, Teleport
+}
+
+struct Token {
+	AssetTransferType type;
+	address token;
+	uint128 amount;
+}
+
+function send(
+  ParaID destinationChain,
+  Beneficiary calldata beneficiary,
+	Token[] tokens,
+  uint128 reward) external payable;
+```
+
+Message enqueued $m_0$:
+
+```solidity
+send(
+	3022, // KILT Para Id
+	Address32(0x0000....),
+	[
+		(ReserveWithdraw, KLT, 100)
+	],
+	10, // WETH
+	)
+```
+
+```solidity
+send { value: 3 }( // Send 3 Eth for fees and reward
+	3022, // KILT Para Id
+	XCMPayload(
+		DepositAsset (KLT, 100) dest=Bob
+		Transact Bob.hello()
+	),
+	[
+		(ReserveWithdraw, KLT, 100)
+	],
+	1, // 1 ETH of 3 needs to be for the reward, the rest is for fees
+	)
+```
+
+The User Agent/UX layer will need to estimate the fee required to be passed into the `send` method. This may be an issue as we cannot Dry-Run something on Polkadot that has not even been submitted on Ethereum yet. We may need to make RPC API to DryRun and get back the xcm that would be submitted to asset hub.
```

Comment:
Why do we need to submit it on Ethereum first? Can't we dry-run the XCM that happens as a result of this command `m`?

Reply:
I think we will need a helper API in our gateway contract that can be used in an off-chain context to produce the  messages/commands that can be used by the dry-run API on BH.